include/vcpkg/tools.h

```diff
@@ -37,6 +39,29 @@ namespace vcpkg
         static constexpr StringLiteral SEVEN_ZIP_MSI = "7zip_msi";
     }
 
+    struct ToolVersion
+    {
+        std::string original_text;
+        std::vector<uint64_t> version;
+
+        friend bool operator==(const ToolVersion& lhs, const ToolVersion& rhs);
+        friend bool operator!=(const ToolVersion& lhs, const ToolVersion& rhs) { return !(lhs == rhs); }
+        friend bool operator<(const ToolVersion& lhs, const ToolVersion& rhs);
+        friend bool operator>(const ToolVersion& lhs, const ToolVersion& rhs) { return rhs < lhs; }
+        friend bool operator>=(const ToolVersion& lhs, const ToolVersion& rhs) { return !(lhs < rhs); }
+        friend bool operator<=(const ToolVersion& lhs, const ToolVersion& rhs) { return !(rhs < lhs); }
+
+        static ToolVersion from_values(std::initializer_list<uint64_t>);
+        static ExpectedL<ToolVersion> try_parse_git(StringView git_version);
+        static ExpectedL<ToolVersion> try_parse_numeric(StringView string_version);
```

Comment:
It is _extremely_ convenient to be able to treat tools generically.

How do you feel about merging these two functions together and activating the git code path iff the version text is prefixed with "git version "?

Reply:
The thing is, we don't treat tools generically. There's always stuff in front to extract out the actual version from what the program spews, and to understand what command line options to supply, etc. Git is different because they mess with the inside of the version number so that goes down to the numeric parsing, but that doesn't seem very different in character to the "provider" machinery we have.

I could see an argument to change the other parser's `\d+(\.\d+(\.\d+))` behavior to look for `\d+(\.\d+|\S+)*` (and strip out the `\S` matches) instead which should work for both forms, but I'm nervous to apply that to all existing tools given that we have relatively poor test coverage here.

Reply:
I was hoping we could extract more things out into common algorithms driven by datatables, but I suppose the need to specify the version query command line (e.g. `git --version` versus `git -v` versus `git help` versus `git`) makes that much less valuable. :(